### ISSUE TITLE
Using unexpected methods, due to dependency conflicts on de.undercouch:bson4jackson:jar

### ISSUE THREAD

**Issue by HelloCoCooo** [restx/restx#297]
Hi, in **restx-restx-parent-0.35.1 (restx-jongo module)**, there are mulptiple versions of **de.undercouch:bson4jackson:jar**. However, according to Maven's _"nearest wins"_ strategy, only **de.undercouch:bson4jackson:jar:2.3.1** can be loaded, and **de.undercouch:bson4jackson:jar:2.4.0** will be shadowed.

As shown in the following dependency tree, **org.jongo:jongo:jar:1.1** expects to reference **de.undercouch:bson4jackson:jar:2.4.0**. But due to dependency conflicts, Maven actually loads **de.undercouch:bson4jackson:jar:2.3.1**. As a result, **org.jongo:jongo:jar:1.1** has to invoke the methods included in the unexpected version **de.undercouch:bson4jackson:jar:2.3.1**, which may cause inconsistent semantic behaviors.

For instance, method **<restx.jongo.Jongos$1: java.lang.Object map(com.mongodb.DBObject)>** actually references method **<de.undercouch.bson4jackson.BsonGenerator: void writeNumber(java.math.BigDecimal)>** in the unexpected version **de.undercouch:bson4jackson:jar:2.3.1** via the following invocation path:

```
<restx.jongo.Jongos$1: java.lang.Object map(com.mongodb.DBObject)> C:\Users\Flipped\Desktop\project\restx-jongo\.\target\classes
<org.jongo.Insert$LazyIdDBObject: java.lang.Object get(java.lang.String)> C:\Users\Flipped\.m2\repository\org\jongo\jongo\1.1\jongo-1.1.jar
<org.jongo.Insert: org.jongo.bson.BsonDocument access$100(org.jongo.marshall.Marshaller,java.lang.Object)> C:\Users\Flipped\.m2\repository\org\jongo\jongo\1.1\jongo-1.1.jar
<org.jongo.Insert: org.jongo.bson.BsonDocument asBsonDocument(org.jongo.marshall.Marshaller,java.lang.Object)> C:\Users\Flipped\.m2\repository\org\jongo\jongo\1.1\jongo-1.1.jar
<org.jongo.marshall.jackson.JacksonEngine: org.jongo.bson.BsonDocument marshall(java.lang.Object)> C:\Users\Flipped\.m2\repository\org\jongo\jongo\1.1\jongo-1.1.jar
<com.fasterxml.jackson.databind.ObjectWriter: void writeValue(java.io.OutputStream,java.lang.Object)> C:\Users\Flipped\.m2\repository\com\fasterxml\jackson\core\jackson-databind\2.8.10\jackson-databind-2.8.10.jar
<com.fasterxml.jackson.databind.ObjectWriter: void _configAndWriteValue(com.fasterxml.jackson.core.JsonGenerator,java.lang.Object)> C:\Users\Flipped\.m2\repository\com\fasterxml\jackson\core\jackson-databind\2.8.10\jackson-databind-2.8.10.jar
<com.fasterxml.jackson.databind.ObjectWriter$Prefetch: void serialize(com.fasterxml.jackson.core.JsonGenerator,java.lang.Object,com.fasterxml.jackson.databind.ser.DefaultSerializerProvider)> C:\Users\Flipped\.m2\repository\com\fasterxml\jackson\core\jackson-databind\2.8.10\jackson-databind-2.8.10.jar
<com.fasterxml.jackson.databind.ser.DefaultSerializerProvider: void serializeValue(com.fasterxml.jackson.core.JsonGenerator,java.lang.Object,com.fasterxml.jackson.databind.JavaType,com.fasterxml.jackson.databind.JsonSerializer)> C:\Users\Flipped\.m2\repository\com\fasterxml\jackson\core\jackson-databind\2.8.10\jackson-databind-2.8.10.jar
<com.fasterxml.jackson.databind.ser.std.NumberSerializer: void serialize(java.lang.Object,com.fasterxml.jackson.core.JsonGenerator,com.fasterxml.jackson.databind.SerializerProvider)> C:\Users\Flipped\.m2\repository\com\fasterxml\jackson\core\jackson-databind\2.8.10\jackson-databind-2.8.10.jar
<com.fasterxml.jackson.databind.ser.std.NumberSerializer: void serialize(java.lang.Number,com.fasterxml.jackson.core.JsonGenerator,com.fasterxml.jackson.databind.SerializerProvider)> C:\Users\Flipped\.m2\repository\com\fasterxml\jackson\core\jackson-databind\2.8.10\jackson-databind-2.8.10.jar
<de.undercouch.bson4jackson.BsonGenerator: void writeNumber(java.math.BigDecimal)>
```

By further analyzing, the expected callee **<de.undercouch.bson4jackson.BsonGenerator: void writeNumber(java.math.BigDecimal)>**, have different implementations from the actual callees with the same signatures (same method names, same paremeters) included in the unexpected (but actual loaded) version **de.undercouch:bson4jackson:jar:2.3.1**, which leads to different behaviors.

**Solution:**
Use the newer version **de.undercouch:bson4jackson:jar:2.4.0** to keep the version consistency.

**Dependency tree----**
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ restx-jongo ---
[INFO] io.restx:restx-jongo:jar:0.35.1
[INFO] +- io.restx:restx-factory:jar:0.35.1:compile
[INFO] |  +- io.restx:restx-common:jar:0.35.1:compile
[INFO] |  |  +- (com.google.guava:guava:jar:18.0:compile - version managed from 11.0.2; omitted for duplicate)
[INFO] |  |  +- (joda-time:joda-time:jar:2.3:compile - omitted for duplicate)
[INFO] |  |  +- (org.reflections:reflections:jar:0.9.9-RC1:compile - omitted for duplicate)
[INFO] |  |  \- com.samskivert:jmustache:jar:1.8:compile
[INFO] |  +- javax.inject:javax.inject:jar:1:compile
[INFO] |  +- com.google.guava:guava:jar:18.0:compile
[INFO] |  +- joda-time:joda-time:jar:2.3:compile
[INFO] |  +- (org.slf4j:slf4j-api:jar:1.7.5:compile - omitted for duplicate)
[INFO] |  \- org.reflections:reflections:jar:0.9.9-RC1:compile
[INFO] |     +- (com.google.guava:guava:jar:18.0:compile - version managed from 11.0.2; omitted for duplicate)
[INFO] |     +- org.javassist:javassist:jar:3.16.1-GA:compile
[INFO] |     \- dom4j:dom4j:jar:1.6.1:compile
[INFO] |        \- xml-apis:xml-apis:jar:1.0.b2:compile
[INFO] +- io.restx:restx-core:jar:0.35.1:compile
[INFO] |  +- (com.fasterxml.jackson.core:jackson-core:jar:2.8.10:compile - omitted for duplicate)
[INFO] |  +- (com.fasterxml.jackson.core:jackson-annotations:jar:2.8.10:compile - version managed from 2.8.0; omitted for duplicate)
[INFO] |  +- (com.fasterxml.jackson.core:jackson-databind:jar:2.8.10:compile - omitted for duplicate)
[INFO] |  +- (com.fasterxml.jackson.datatype:jackson-datatype-joda:jar:2.8.10:compile - omitted for duplicate)
[INFO] |  +- com.fasterxml.jackson.datatype:jackson-datatype-guava:jar:2.8.10:compile
[INFO] |  |  +- (com.fasterxml.jackson.core:jackson-databind:jar:2.8.10:compile - omitted for duplicate)
[INFO] |  |  +- (com.fasterxml.jackson.core:jackson-core:jar:2.8.10:compile - omitted for duplicate)
[INFO] |  |  \- (com.google.guava:guava:jar:18.0:compile - version managed from 16.0; omitted for duplicate)
[INFO] |  +- (io.restx:restx-factory:jar:0.35.1:compile - omitted for duplicate)
[INFO] |  +- io.restx:restx-classloader:jar:0.35.1:compile
[INFO] |  |  +- (io.restx:restx-common:jar:0.35.1:compile - omitted for duplicate)
[INFO] |  |  +- (javax.inject:javax.inject:jar:1:compile - omitted for duplicate)
[INFO] |  |  +- (com.google.guava:guava:jar:18.0:compile - version managed from 16.0; omitted for duplicate)
[INFO] |  |  +- (joda-time:joda-time:jar:2.3:compile - version managed from 2.7; omitted for duplicate)
[INFO] |  |  \- (org.slf4j:slf4j-api:jar:1.7.5:compile - omitted for duplicate)
[INFO] |  +- io.restx:restx-apidocs-doclet:jar:0.35.1:compile
[INFO] |  |  +- com.sun.tools:tools:jar:1.8.0_191:system
[INFO] |  |  +- (com.fasterxml.jackson.core:jackson-core:jar:2.8.10:compile - omitted for duplicate)
[INFO] |  |  +- (com.fasterxml.jackson.core:jackson-annotations:jar:2.8.10:compile - version managed from 2.8.0; omitted for duplicate)
[INFO] |  |  +- (com.fasterxml.jackson.core:jackson-databind:jar:2.8.10:compile - omitted for duplicate)
[INFO] |  |  +- (com.google.guava:guava:jar:18.0:compile - version managed from 16.0; omitted for duplicate)
[INFO] |  |  \- (joda-time:joda-time:jar:2.3:compile - version managed from 2.7; omitted for duplicate)
[INFO] |  +- (javax.inject:javax.inject:jar:1:compile - omitted for duplicate)
[INFO] |  +- (com.google.guava:guava:jar:18.0:compile - version managed from 16.0; omitted for duplicate)
[INFO] |  +- (joda-time:joda-time:jar:2.3:compile - version managed from 2.7; omitted for duplicate)
[INFO] |  +- (org.slf4j:slf4j-api:jar:1.7.5:compile - omitted for duplicate)
[INFO] |  +- javax.validation:validation-api:jar:1.1.0.Final:compile
[INFO] |  +- (org.reflections:reflections:jar:0.9.9-RC1:compile - omitted for duplicate)
[INFO] |  \- org.yaml:snakeyaml:jar:1.13:compile
[INFO] +- io.restx:restx-core-annotation-processor:jar:0.35.1:compile
[INFO] |  +- (io.restx:restx-factory:jar:0.35.1:compile - omitted for duplicate)
[INFO] |  \- (io.restx:restx-core:jar:0.35.1:compile - omitted for duplicate)
[INFO] +- io.restx:restx-security-basic:jar:0.35.1:compile
[INFO] |  +- (io.restx:restx-factory:jar:0.35.1:compile - omitted for duplicate)
[INFO] |  +- (io.restx:restx-core:jar:0.35.1:compile - omitted for duplicate)
[INFO] |  +- (io.restx:restx-core-annotation-processor:jar:0.35.1:compile - omitted for duplicate)
[INFO] |  \- org.mindrot:jbcrypt:jar:0.3m:compile
[INFO] +- com.fasterxml.jackson.core:jackson-core:jar:2.8.10:compile
[INFO] +- com.fasterxml.jackson.core:jackson-annotations:jar:2.8.10:compile
[INFO] +- com.fasterxml.jackson.core:jackson-databind:jar:2.8.10:compile
[INFO] |  +- (com.fasterxml.jackson.core:jackson-annotations:jar:2.8.10:compile - version managed from 2.8.0; omitted for duplicate)
[INFO] |  \- (com.fasterxml.jackson.core:jackson-core:jar:2.8.10:compile - omitted for duplicate)
[INFO] +- com.fasterxml.jackson.datatype:jackson-datatype-joda:jar:2.8.10:compile
[INFO] |  +- (com.fasterxml.jackson.core:jackson-annotations:jar:2.8.10:compile - version managed from 2.8.0; omitted for duplicate)
[INFO] |  +- (com.fasterxml.jackson.core:jackson-core:jar:2.8.10:compile - omitted for duplicate)
[INFO] |  +- (com.fasterxml.jackson.core:jackson-databind:jar:2.8.10:compile - omitted for duplicate)
[INFO] |  \- (joda-time:joda-time:jar:2.3:compile - version managed from 2.7; omitted for duplicate)
[INFO] +- **de.undercouch:bson4jackson:jar:2.3.1:compile**
[INFO] |  +- (com.fasterxml.jackson.core:jackson-databind:jar:2.8.10:compile - version managed from 2.3.0; omitted for duplicate)
[INFO] |  +- (com.fasterxml.jackson.core:jackson-core:jar:2.8.10:compile - version managed from 2.3.0; omitted for duplicate)
[INFO] |  \- (com.fasterxml.jackson.core:jackson-annotations:jar:2.8.10:compile - version managed from 2.3.0; omitted for duplicate)
[INFO] +- org.mongodb:mongo-java-driver:jar:2.11.3:compile
[INFO] +- **org.jongo:jongo:jar:1.1:compile**
[INFO] |  +- (com.fasterxml.jackson.core:jackson-core:jar:2.8.10:compile - version managed from 2.4.1; omitted for duplicate)
[INFO] |  +- (com.fasterxml.jackson.core:jackson-databind:jar:2.8.10:compile - version managed from 2.4.1; omitted for duplicate)
[INFO] |  +- (com.fasterxml.jackson.core:jackson-annotations:jar:2.8.10:compile - version managed from 2.4.1; omitted for duplicate)
[INFO] |  \- **(de.undercouch:bson4jackson:jar:2.3.1:compile - version managed from 2.4.0; omitted for duplicate)**
[INFO] +- org.slf4j:slf4j-api:jar:1.7.5:compile
[INFO] +- junit:junit:jar:4.11:test
[INFO] |  \- org.hamcrest:hamcrest-core:jar:1.3:test
[INFO] \- org.assertj:assertj-core:jar:1.6.0:test

Thanks!
Best regards,
Coco

**Comment by HelloCoCooo** [comment]
Code snippet of **<de.undercouch.bson4jackson.BsonGenerator: void writeNumber(java.math.BigDecimal)>** in **de.undercouch:bson4jackson:jar:2.3.1:compile**:
![write2 3 1](https://user-images.githubusercontent.com/46306510/53284182-8736ad00-378b-11e9-82aa-82e7b09cc0d6.png)


Code snippet of **<de.undercouch.bson4jackson.BsonGenerator: void writeNumber(java.math.BigDecimal)>** in **de.undercouch:bson4jackson:jar:2.4.0:compile**:
![write2 4 0](https://user-images.githubusercontent.com/46306510/53284184-8e5dbb00-378b-11e9-9f37-71728ac28929.png)

As shown in the comments, method **<de.undercouch.bson4jackson.BsonGenerator: void writeNumber(java.math.BigDecimal)>** included in newer version **cde.undercouch:bson4jackson:jar:2.4.0** deals with more cases and adds more control branches, which changes the control flows and data flows. So being forced to use older version **de.undercouch:bson4jackson:jar:2.3.1** may lead to inconsisitent semantic behaviors.

**Comment by HelloCoCooo** [comment]
@fcamblor May I pull a request to fix this issue?

**Comment by fcamblor** [comment]
Hi @HelloCoCooo and thanks for the detailed feedback.

Let me have a look at it more precisely as I know I had to fight against bson4jackson inconsistencies in the past, I need to get back the history in my head prior to tell you if you can go ahead with a PR (and if the way you're going to propose with match with my previous issues :-) )

I'll have a look at this just now.

**Comment by fcamblor** [comment]
BTW can I ask you to paste your current pom's dependency declaration (order & scopes) ?

I know I should be able to deduct it from the `dependency:tree` call but it would be more readable to only see "root" dependencies from the tree.

**Comment by fcamblor** [comment]
OK so to summarize, it seems like it's due to `jongo:1.0` => `jongo:1.1` upgrade between `restx:0.34` and `restx:0.35.1`

On `restx:0.34`, we had something like :
```
restx-jongo:0.34
+- de.undercouch:bson4jackson:jar:2.3.1:compile
+- org.jongo:jongo:jar:1.0:compile
|  \- (de.undercouch:bson4jackson:jar:2.3.1:compile - version managed from 2.2.3; omitted for duplicate)
```

On `restx:0.35.1` :
```
+- de.undercouch:bson4jackson:jar:2.3.1:compile
+- org.jongo:jongo:jar:1.1:compile
|  \- (de.undercouch:bson4jackson:jar:2.3.1:compile - version managed from 2.4.0; omitted for duplicate)
```

> **Solution:**
>  Use the newer version **de.undercouch:bson4jackson:jar:2.4.0** to keep the version consistency.

My view on this would be to rather remove the explicit `restx-jongo`'s `de.undercouch:bson4jackson` dependency, relying on what is pulled transitively by `jongo`
That way, we would avoid such inconsistencies in the future, WDYT ?

**Comment by fcamblor** [comment]
closed through #298 thanks @HelloCoCooo
